pkg/controllers/raycluster_webhook_test.go

```diff
@@ -0,0 +1,397 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"testing"
+
+	"github.com/onsi/gomega"
+	. "github.com/project-codeflare/codeflare-common/support"
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/project-codeflare/codeflare-operator/pkg/config"
+)
+
+var (
+	namespace = "test-namespace"
+
+	rcWebhook = &rayClusterWebhook{
+		Config: &config.KubeRayConfiguration{},
+	}
+)
+
+func TestRayClusterWebhookDefault(t *testing.T) {
+	test := NewTest(t)
+
+	validRayCluster := &rayv1.RayCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-valid-raycluster",
+			Namespace: namespace,
+		},
+		Spec: rayv1.RayClusterSpec{
+			HeadGroupSpec: rayv1.HeadGroupSpec{
+				Template: corev1.PodTemplateSpec{
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{},
+					},
+				},
+				RayStartParams: map[string]string{},
+			},
+			WorkerGroupSpecs: []rayv1.WorkerGroupSpec{
+				{
+					GroupName: "worker-group-1",
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Name: "worker-container-1",
+								},
+							},
+						},
+					},
+					RayStartParams: map[string]string{},
+				},
+				{
+					GroupName: "worker-group-2",
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Name: "worker-container-2",
+								},
+							},
+						},
+					},
+					RayStartParams: map[string]string{},
+				},
+			},
+		},
+	}
+
+	// Create the RayClusters
+	if _, err := test.Client().Ray().RayV1().RayClusters(namespace).Create(test.Ctx(), validRayCluster, metav1.CreateOptions{}); err != nil {
+		test.T().Fatalf("Failed to create RayCluster: %v", err)
+	}
+
+	// Call to default function is made
+	err := rcWebhook.Default(test.Ctx(), runtime.Object(validRayCluster))
+	t.Run("Expected no errors on call to Default function", func(t *testing.T) {
+		g := gomega.NewWithT(t)
+		g.Expect(err).To(gomega.BeNil())
+	})
+
+	t.Run("Expected required OAuth proxy container for the head group", func(t *testing.T) {
+		g := gomega.NewWithT(t)
+		g.Expect(func() bool {
+			for _, container := range validRayCluster.Spec.HeadGroupSpec.Template.Spec.Containers {
+				if container.Name == oauthProxyContainerName {
+					return true
+				}
+			}
+			return false
+		}()).To(gomega.BeTrue(), "Expected the OAuth proxy container to be present in the list of head group containers")
+	})
```

Comment:
This boilerplate isn't needed, you can leverage `test` directly as it embeds gomega:

```
	test.Expect(err).NotTo(HaveOccurred(), "Expected no errors on call to Default function")
	test.Expect(validRayCluster.Spec.HeadGroupSpec.Template.Spec.Containers).
		To(ContainElement(WithTransform(ContainerName, Equal(oauthProxyContainerName))), "Expected the OAuth proxy container to be present in the list of head group containers")
```

With helper functions in the bottom (to be relocated into codeflare-common):
```
func ContainerName(container corev1.Container) string {
	return container.Name
}
```

Reply:
Also this approach avoids `BeTrue()` checks for element with specific value, which often doesn't provide meaningful context.

Reply:
Test can also check number of elements, making the assertion more structured:
```
	test.Expect(validRayCluster.Spec.HeadGroupSpec.Template.Spec.Containers).
		To(
			And(
				HaveLen(1),
				ContainElement(WithTransform(ContainerName, Equal(oauthProxyContainerName))),
			),
			"Expected the OAuth proxy container to be present in the list of head group containers",
		)
```